### PR TITLE
Add MPI support to ParaGrid

### LINKED ISSUE #448: Move MPI Parallelisation to ParGrid and remove RectGrid
Following the discussion in #442 , we need to move the MPI parallelisation done in #331 over to the parametric grid, and then remove the rectangular grid.

This should be done after #331 and #432 were merged into `develop`.

### LINKED ISSUE #454: some tests have been disabled (temporary) whilst we implement new MPI features
2 tests have been temporarily disabled (as part of #432):
* `testParaGrid`
* `testConfigOutput`

I have edited the `CMakeLists.txt` file to print a warning to users at build time.
```
CMake Warning at core/test/CMakeLists.txt:57 (message):
  testParaGrid has been temporarily disabled when running with MPI enabled

CMake Warning at core/test/CMakeLists.txt:79 (message):
  testConfigOutput has been temporarily disabled when running with MPI enabled
```

They have been disabled because currently ParaGrid has not implemented MPI. Therefore, building against MPI-enabled `nextsimlib` will cause the tests to fail. 

Once @MarionBWeinzierl has finished adding MPI to Paragrid we can re-enable them.

### LINKED ISSUE #534: single column test fails with MPI build on
If you build `nextsim` with `MPI=ON`, and run the `ThermoIntegration_test.py` integration test it will produce the following error (on Linux):
<details><summary>Error Output</summary>

```
$ python ThermoIntegration_test.py 
HDF5-DIAG: Error detected in HDF5 (1.8.21) thread 0:
  #000: /tmp/melt/spack-stage/spack-stage-hdf5-1.8.21-dfgfamgva4yyx72fmjpldk33s3e6uap6/spack-src/src/H5A.c line 1638 in H5Aexists(): not a location
    major: Invalid arguments to routine
    minor: Inappropriate type
  #001: /tmp/melt/spack-stage/spack-stage-hdf5-1.8.21-dfgfamgva4yyx72fmjpldk33s3e6uap6/spack-src/src/H5Gloc.c line 193 in H5G_loc(): invalid group ID
    major: Invalid arguments to routine
    minor: Bad value
HDF5-DIAG: Error detected in HDF5 (1.8.21) thread 0:
  #000: /tmp/melt/spack-stage/spack-stage-hdf5-1.8.21-dfgfamgva4yyx72fmjpldk33s3e6uap6/spack-src/src/H5Adeprec.c line 176 in H5Acreate1(): not a location
    major: Invalid arguments to routine
    minor: Inappropriate type
  #001: /tmp/melt/spack-stage/spack-stage-hdf5-1.8.21-dfgfamgva4yyx72fmjpldk33s3e6uap6/spack-src/src/H5Gloc.c line 193 in H5G_loc(): invalid group ID
    major: Invalid arguments to routine
    minor: Bad value
terminate called after throwing an instance of 'netCDF::exceptions::NcFileMeta'
  what():  NetCDF: Can't add HDF5 file metadata
file: ncFile.cpp  line:33
[lenny:23295] *** Process received signal ***
[lenny:23295] Signal: Aborted (6)
[lenny:23295] Signal code:  (-6)
[lenny:23295] [ 0] /lib/x86_64-linux-gnu/libc.so.6(+0x42520)[0x7d9550bec520]
[lenny:23295] [ 1] /lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x12c)[0x7d9550c409fc]
[lenny:23295] [ 2] /lib/x86_64-linux-gnu/libc.so.6(raise+0x16)[0x7d9550bec476]
[lenny:23295] [ 3] /lib/x86_64-linux-gnu/libc.so.6(abort+0xd3)[0x7d9550bd27f3]
[lenny:23295] [ 4] /lib/x86_64-linux-gnu/libstdc++.so.6(+0xa2b9e)[0x7d9550e95b9e]
[lenny:23295] [ 5] /lib/x86_64-linux-gnu/libstdc++.so.6(+0xae20c)[0x7d9550ea120c]
[lenny:23295] [ 6] /lib/x86_64-linux-gnu/libstdc++.so.6(+0xae277)[0x7d9550ea1277]
[lenny:23295] [ 7] /lib/x86_64-linux-gnu/libstdc++.so.6(+0xae4d8)[0x7d9550ea14d8]
[lenny:23295] [ 8] /software/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/netcdf-cxx4-4.3.1-kwvtkd7klice2xvellynedzj63eraxly/lib/libnetcdf_c++4.so.1(+0x26a2a)[0x7d9550aa3a2a]
[lenny:23295] [ 9] /software/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.4.0/netcdf-cxx4-4.3.1-kwvtkd7klice2xvellynedzj63eraxly/lib/libnetcdf_c++4.so.1(_ZN6netCDF6NcFile5closeEv+0x33)[0x7d9550aab123]
[lenny:23295] [10] /home/melt/sync/cambridge/projects/current/sasip/nextsimdg/build-mpi/libnextsimlib.so(_ZN7Nextsim10ParaGridIO5closeERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x4e)[0x7d955276f43e]
[lenny:23295] [11] /home/melt/sync/cambridge/projects/current/sasip/nextsimdg/build-mpi/libnextsimlib.so(_ZN7Nextsim10ParaGridIO13closeAllFilesEv+0x9f)[0x7d955276f50d]
[lenny:23295] [12] /lib/x86_64-linux-gnu/libc.so.6(+0x45495)[0x7d9550bef495]
[lenny:23295] [13] /lib/x86_64-linux-gnu/libc.so.6(on_exit+0x0)[0x7d9550bef610]
[lenny:23295] [14] /lib/x86_64-linux-gnu/libc.so.6(+0x29d97)[0x7d9550bd3d97]
[lenny:23295] [15] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7d9550bd3e40]
[lenny:23295] [16] ../nextsim(+0xc5f5)[0x5a2b7ab945f5]
[lenny:23295] *** End of error message ***
Aborted (core dumped)
E
======================================================================
ERROR: setUpClass (__main__.SingleColumnThermo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/melt/sync/cambridge/projects/current/sasip/nextsimdg/test/ThermoIntegration_test.py", line 38, in setUpClass
    subprocess.run(cls.executable + " --config-file " + cls.config_file, shell=True, check=True)
  File "/home/melt/miniconda3/envs/nextsim/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '../nextsim --config-file ThermoIntegration.cfg' returned non-zero exit status 134.

----------------------------------------------------------------------
Ran 0 tests in 45.091s

FAILED (errors=1)
```
</details> 

I am investigating why exactly this happens.

The single column test runs when `MPI=OFF`

### PR DESCRIPTION
**Pull request by TomMelt**
# Add MPI support to ParaGrid
closes #534
closes #454
closes #448 

---
# Task List
- [x] modify `make_init_para24x30.py` to use new DG data layout and remove CG vars
- [x] add MPI doctest support to `ParaGrid_test.cpp`
- [x] move `extent` and `start` variables into `ModelArray`'s `DimensionSpec` 
- [x] add special MPI test case to `ParaGrid_test.cpp`
- [x] add MPI support to `getModelState`
- [x] add MPI support to `dumpModelState`
- [x] add MPI support to `writeDiagnosticTime`
- [x] reinstate test `ConfigOutput_test.cpp` for MPI builds
- [x] add MPI support to `ERA5`/`TOPAZ` tests

---
# Change Description

After #331 added MPI parallelisation for thermodynamics on the RectGrid, this PR does the same for the Parametric grid. This should then check off the second task in #120 (_"MPI parallelization of thermodynamics where all operations, except for I/O, are local to an MPI rank"_)

---
# Test Description

- `ParaGrid_test.cpp` tests core functionality of ParaGrid (serial and MPI)
- `./nextsim --config-file config_para24x30.cfg` should provide an integration test (serial and MPI) (based on #506)

---
# Further work (for a future PR)

- add MPI to dynamics (to close PR #120)
- implement halo comms
- implement boundary conditions (as part of MPI) 
- tidy naming of variables in Domain Decomp tool

### CODE REVIEW

File: core/test/ParaGrid_test.cpp [hunk 1]
```diff
@@ -161,18 +207,45 @@ TEST_CASE("Write and read a ModelState-based ParaGrid restart file")
...
+        metadata.globalExtentX = nx;
+        metadata.globalExtentY = ny;
+        metadata.localCornerX = start_x;
+        metadata.localCornerY = 0;
+        metadata.localExtentX = local_nx;
+        metadata.localExtentY = ny;
+    }
+    if (metadata.mpiMyRank == 1) {
+        metadata.setMpiMetadata(test_comm);
+        metadata.globalExtentX = local_nx;
```
Comment (jwallwork23) [inline core/test/ParaGrid_test.cpp:223]: Is this right?

File: core/test/ParaGrid_test.cpp [hunk 2]
```diff
@@ -342,21 +407,46 @@ TEST_CASE("Write a diagnostic ParaGrid file")
...
```
Comment (jwallwork23) [inline core/test/ParaGrid_test.cpp:431]: Again, is this right?

File: core/test/ParaGrid_test.cpp [hunk 3]
```diff
@@ -420,28 +515,39 @@ TEST_CASE("Test array ordering")
...
 
     HField index2d(ModelArray::Type::H);
     index2d.resize();
     std::string fieldName = "index2d";
     std::set<std::string> fields = { fieldName };
     TimePoint time;
 
-    ModelState state = ParaGridIO::readForcingTimeStatic(
-        fields, time, TO_STR(TEST_FILE_SOURCE) + std::string("/") + inputFilename);
+    printf("path = %s\n", inputFilename.c_str());
```
Comment (jwallwork23) [inline core/test/ParaGrid_test.cpp:539]: How come you used `printf` rather than `std::cout`?

**Review by jwallwork23 — COMMENTED** [review]
Thanks @TomMelt! Some initial comments.

File: core/test/ParaGrid_test.cpp (hunk 3 above)
Comment (TomMelt) [reply core/test/ParaGrid_test.cpp]: I didn't mean to leave this in anyway. It's been removed. It was for debugging purposes.

on a side note though... I use `printf` because it doesn't require additional header files.

File: core/src/ParaGridIO.cpp [hunk 4]
```diff
@@ -72,31 +76,72 @@ void ParaGridIO::makeDimCompMap()
...
                 // the number that the selected ice thermodynamics requires.
-                ModelArray::setDimension(entry.first, NZLevels::get());
+#ifdef USE_MPI
+                ModelArray::setDimension(dimType, NZLevels::get(), NZLevels::get(), 0);
+#else
+                ModelArray::setDimension(dimType, NZLevels::get());
+#endif
             } else {
-                ModelArray::setDimension(entry.first, dim.getSize());
+              // this needs MPIifying
```
[the remark this answers is not in the document]
Comment (TomMelt) [reply core/src/ParaGridIO.cpp]: sorry this has been removed

File: core/test/ParaGrid_test.cpp (hunk 1 above)
Comment (TomMelt) [reply core/test/ParaGrid_test.cpp]: thanks for spotting this. It's been fixed :+1:

File: core/src/include/ModelArray.hpp [hunk 5]
```diff
@@ -272,15 +274,19 @@ class ModelArray {
      * @param dim The dimension to be altered.
      * @param length The new length of the dimension.
      */
-    static void setDimension(Dimension dim, size_t length);
+#ifdef USE_MPI
+    static void setDimension(Dimension dim, size_t global_length, size_t local_length, size_t size);
+#else
+    static void setDimension(Dimension dim, size_t global_length);
+#endif
```
[the remark this answers is not in the document]
Comment (TomMelt) [reply core/src/include/ModelArray.hpp:282]: I did think about this, but it's much nicer to get compile time warning (in my opinion). Mostly because it stung me a few times already :sweat:  and I think it could catch future devs out as well.

Happy to discuss in dev meeting though as I don't have a strong opinion either way. I did initially code as optional but I removed that when it prevented me finding bugs.

[the remark this answers is not in the document]
Comment (TomMelt) [reply core/src/include/ModelArray.hpp:282]: tldr; using the optional approach, it is possible to use the wrong function call and you will only find out at run time.

File: core/src/include/ParaGridIO.hpp [hunk 6]
```diff
@@ -35,7 +35,11 @@ class ParaGridIO : public ParametricGrid::IParaGridIO {
      * Retrieves the ModelState from a restart file of the parametric_grid type.
      * @param filePath The file path containing the file to be read.
      */
+#ifdef USE_MPI
+    ModelState getModelState(const std::string& filePath, ModelMetadata& metadata) override;
+#else
     ModelState getModelState(const std::string& filePath) override;
+#endif
```
[the remark this answers is not in the document]
Comment (TomMelt) [reply core/src/include/ParaGridIO.hpp:42]: Thanks for raising. My prev point also works here. We can discuss in the team meeting :+1:

[the remark this answers is not in the document]
Comment (timspainNERSC) [reply core/src/include/ParaGridIO.hpp:42]: There's an argument for passing `ModelMetadata` to `getModelState()`. It makes the pair of functions `getModelState()` and `dumpModelState()` more symmetrical and perhaps some of the logic for setting the model state would be better in the `getModelState()` function?

[the remark this answers is not in the document]
Comment (TomMelt) [reply core/src/include/ParaGridIO.hpp:42]: > There's an argument for passing ModelMetadata to getModelState(). It makes the pair of functions getModelState() and dumpModelState() more symmetrical and perhaps some of the logic for setting the model state would be better in the getModelState() function?

I think we should raise this as a separate issue (with accompanying PR)

File: core/src/ModelArray.cpp [hunk 7]
```diff
@@ -200,9 +200,23 @@ void ModelArray::setNComponents(std::map<Type, size_t> cMap)
...
+#ifdef USE_MPI
+    definedDimensions.at(dim).global_length = global_length;
+    definedDimensions.at(dim).local_length = local_length;
+    definedDimensions.at(dim).start = start;
+#else
+    // if MPI is not used then set the local_length to be the same as the global
+    definedDimensions.at(dim).global_length = global_length;
+    definedDimensions.at(dim).local_length = global_length;
+    definedDimensions.at(dim).start = 0;
+#endif
```
Comment (timspainNERSC) [inline core/src/ModelArray.cpp:219]: Rewrite this to only require one `#ifdef`

File: core/test/ConfigOutput_test.cpp [hunk 8]
```diff
@@ -79,21 +103,29 @@ TEST_CASE("Test periodic output")
...
     }
     for (size_t j = 0; j < ny; ++j) {
-        for (size_t i = 0; i < nx; ++i) {
-            hice(i, j) = 0 + 0.01 * (j * nx + i);
-            cice(i, j) = 0.1 + 0.01 * (j * nx + i);
-            hsnow(i, j) = 0.2 + 0.01 * (j * nx + i);
+        for (size_t i = 0; i < local_nx; ++i) {
+            hice(i, j) = 0 + 0.01 * (j * nx + (i + start_x));
+            cice(i, j) = 0.1 + 0.01 * (j * nx + (i + start_x));
+            hsnow(i, j) = 0.2 + 0.01 * (j * nx + (i + start_x));
```
Comment (timspainNERSC) [inline core/test/ConfigOutput_test.cpp:128]: Variable names should be CamelCased

File: core/test/ParaGrid_test.cpp [hunk 9]
```diff
@@ -26,19 +32,67 @@
...
 #include <ncVar.h>
 
-const std::string filename = "paraGrid_test.nc";
+const std::string test_files_dir = TEST_FILES_DIR;
+const std::string filename = test_files_dir + "/paraGrid_test.nc";
 const std::string diagFile = "paraGrid_diag.nc";
+const std::string date_string = "2000-01-01T00:00:00Z";
+#ifdef USE_MPI
+const std::string partition_filename = test_files_dir + "/partition_metadata_2.nc";
+#endif
```
Comment (timspainNERSC) [inline core/test/ParaGrid_test.cpp:41]: Variable names should be CamelCased.

File: core/test/ParaGrid_test.cpp [hunk 10]
```diff
@@ -74,22 +133,13 @@ TEST_CASE("Write and read a ModelState-based ParaGrid restart file")
...
-        for (size_t i = 0; i < nx; ++i) {
-            fractional(i, j) = j * yFactor + i * xFactor;
-            mask(i, j)
-                = (i - nx / 2) * (i - nx / 2) + (j - ny / 2) * (j - ny / 2) > (nx * ny) ? 0 : 1;
-            for (size_t d = 0; d < DG; ++d) {
-                fractionalDG.components({ i, j })[d] = fractional(i, j) + d;
-            }
-        }
-    }
+    initialize_test_data(fractional, fractionalDG, mask);
```
Comment (timspainNERSC) [inline core/test/ParaGrid_test.cpp:136]: Function name should be CamelCased.

File: core/test/ParaGrid_test.cpp [hunk 11]
```diff
@@ -99,34 +149,29 @@ TEST_CASE("Write and read a ModelState-based ParaGrid restart file")
...
-    REQUIRE(coordinates.components({ 12, 13 })[0] - coordinates.components({ 11, 13 })[0] == scale);
-    REQUIRE(coordinates.components({ 12, 13 })[1] - coordinates.components({ 12, 12 })[1] == scale);
+    // MPI domain is only split in x-direction for this test
+    // the following will be set correctly with MPI ON and OFF
+    auto dim_x = ModelArray::Dimension::X;
+    auto start_x = ModelArray::definedDimensions.at(dim_x).start;
+    auto local_nx = ModelArray::definedDimensions.at(dim_x).local_length;
+    auto dim_x_vertex = ModelArray::Dimension::XVERTEX;
+    auto local_nx_vertex = ModelArray::definedDimensions.at(dim_x_vertex).local_length;
+    auto start_x_vertex = ModelArray::definedDimensions.at(dim_x_vertex).start;
```
Comment (timspainNERSC) [inline core/test/ParaGrid_test.cpp:164]: Variable names should be CamelCased.

**Review by timspainNERSC — CHANGES_REQUESTED** [review]

**Comment by timspainNERSC** [comment]
As stated in the README, we follow WebKit style
https://webkit.org/code-style-guidelines/#names-basic

File: core/src/ModelArray.cpp (hunk 7 above)
Comment (TomMelt) [reply core/src/ModelArray.cpp]: done

File: core/src/include/ModelArray.hpp [hunk 12]
```diff
@@ -57,7 +57,9 @@ class ModelArray {
     struct DimensionSpec {
         std::string name;
         std::string altName;
-        size_t length;
+        size_t global_length;
+        size_t local_length;
+        size_t start;
```
[the remark this answers is not in the document]
Comment (TomMelt) [reply core/src/include/ModelArray.hpp:62]: done

File: core/src/include/ModelArray.hpp [hunk 13]
```diff
@@ -57,7 +57,9 @@ class ModelArray {
...
     };
```
[the remark this answers is not in the document]
Comment (TomMelt) [reply core/src/include/ModelArray.hpp]: Do you mean that you want separate setters for each variable e.g., `setDimensionLocalLength`, `setDimensionGlobalLength`?

And by _keeping with the class_ do you mean in the header file i.e., `core/src/include/ModelArray.hpp`?

File: core/test/ParaGrid_test.cpp (hunk 11 above)
Comment (TomMelt) [reply core/test/ParaGrid_test.cpp]: done

File: core/test/ParaGrid_test.cpp (hunk 10 above)
Comment (TomMelt) [reply core/test/ParaGrid_test.cpp]: done

File: core/test/ConfigOutput_test.cpp (hunk 8 above)
Comment (TomMelt) [reply core/test/ConfigOutput_test.cpp]: done

File: core/src/include/ModelArray.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (TomMelt) [reply core/src/include/ModelArray.hpp]: Thanks @timspainNERSC , this has now been resolve by f22b88f3

File: core/src/include/ParaGridIO.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (TomMelt) [reply core/src/include/ParaGridIO.hpp:43]: After discussion today, that we are planning to only support MPI in future, we probably don't need to consider this any further. The serial versions of `dumpModelState` and `getModelState` will removed.

**Review by timspainNERSC — APPROVED** [review]
